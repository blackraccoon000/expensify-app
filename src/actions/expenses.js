import uuid from "uuid";
import database from "../firebase/firebase"

const addExpense = (expense) => ({
  type:"ADD_EXPENSE",
  expense
})

const startAddExpense = (expenseData = {}) => {
  return (dispatch, getState) => {
    // console.log(getState())
    const uid = getState().auth.uid
    const {
      description = "",
      note = "",
      amount = 0,
      createdAt = 0
    } = expenseData
    const expense = { description, note, amount, createdAt }

    return database.ref(`users/${uid}/expenses`).push(expense)
    // return database.ref("expenses").push(expense)
      .then((ref)=>{
        dispatch(addExpense({
          id:ref.key,
          ...expense
        }))
      })
  }
}

const removeExpense = ({id=""}={}) => {
  return {
    type:"REMOVE_EXPENSE",
    id
  }
}

const startRemoveExpense = ({id=""}={}) => {
  return (dispatch, getState) => {
    const uid = getState().auth.uid
    return database.ref(`users/${uid}/expenses/${id}`).remove().then(()=>{
      dispatch(removeExpense({id}))
    })
  }
}

const editExpense = (id, updates) => ({
  type:"EDIT_EXPENSE",
  id,
  updates
})

const startEditExpense = (id, updates) => {
  return (dispatch,getState) => {
    const uid = getState().auth.uid
    return database.ref(`users/${uid}/expenses/${id}`).update(updates)
    .then(()=>{
      dispatch(editExpense(id, updates))
    })
  }
}

const setExpenses = expenses => ({
  type:"SET_EXPENSES",
  expenses
})

const startSetExpenses = () => {
  return (dispatch,getState) => {
    const uid = getState().auth.uid
    return database.ref(`users/${uid}/expenses`).once("value")
    .then(snapshot => {
      const expenses = []
      snapshot.forEach(childSnapshot => {
          expenses.push({
            id: childSnapshot.key,
            ...childSnapshot.val()
          })
        })
      dispatch(setExpenses(expenses))
    })
  }
}



export { addExpense, removeExpense, editExpense, setExpenses,
  startAddExpense, startRemoveExpense, startEditExpense, startSetExpenses }