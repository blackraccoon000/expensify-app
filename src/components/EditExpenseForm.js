import React from "react"
import moment from "moment"

export default class EditExpenseForm extends React.Component {
  constructor(props) {
    super(props)
    this.state = {
      amount: props.amount,
      calenderFocused: false,
      createdAt: moment(),
      description: props.description ,
      error:undefined,
      note: props.note
    }
  }
  onAmountChange = (e) => {
    const amount = e.target.value
    if(!amount || amount.match(/^\d{1,7}(\.\d{0,2})?$/)) {
      this.setState(()=>({ amount }))
    }
  }

  onDescriptionChange = (e) => {
    const description = e.target.value
    this.setState(()=>({ description }))
  }

  onNoteChange = (e) => {
    const note = e.target.value
    this.setState(()=>({ note }))
  }

  onSubmit = (e) => {
    e.preventDefault()
    if(!this.state.description || !this.state.amount) {
      this.setState(()=>{ error:"Please provide description and amount."})
    } else {
      this.setState(()=>{ error:undefined })
      this.props.onSubmit({
        description:this.state.description,
        amount: parseFloat(this.state.amount,10),
        createdAt: this.state.createdAt.valueOf(),
        note:this.state.note
      })
    }
  }

  render() {
    return (
      <div>
        <form onSubmit={this.onSubmit}>
          <label>
            Des:
            <input type="text" value={this.state.description} onChange={this.onDescriptionChange}/>
          </label>
          <label>
            Amount:
            <input type="text" value={this.state.amount} onChange={this.onAmountChange}/>
          </label>
          <label>
            Note:
            <textarea value={this.state.note} onChange={this.onNoteChange}></textarea>
          </label>
          <button>Edit Expense</button>
        </form>
      </div>
    )
  }
}