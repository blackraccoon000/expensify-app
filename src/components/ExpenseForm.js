import React from "react"
import moment from "moment"

// import { connect } from "react-redux"
// import { addExpense } from "../actions/expenses"
import { SingleDatePicker } from "react-dates"

export default class ExpenseForm extends React.Component {
  constructor(props){
    super(props)
    this.state = {
      amount:props.expense ? props.expense.amount.toString() : "",
      calenderFocused: false,
      createdAt: props.expense ? moment(props.expense.createdAt) : moment(),
      description: props.expense ? props.expense.description : "",
      error:undefined,
      note: props.expense ? props.expense.note : "",
    }
  }

  onAmountChange = (e) => {
    const amount = e.target.value

    if(!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/)){
      // console.log(amount)
      this.setState(()=>({ amount }))
    }
  }
  onDateChange = (createdAt) => {
    if(createdAt) {
      this.setState(()=>({ createdAt }))
    }
  }
  onDescriptionChange = (e) => {
    const description = e.target.value
    this.setState(()=>({ description }))
  }
  onFocusChange = ({ focused }) => {
    // console.log(focused)
    this.setState(()=>({ calenderFocused: focused }))
  }
  onNoteChange = (e) => {
    const note = e.target.value
    this.setState(()=>({ note }))
  }
  onSubmit = (e) => {
    e.preventDefault()
    if(!this.state.description || !this.state.amount) {
      this.setState(()=>({ error:"Please provide description and amount." }))
    } else {
      this.setState(()=>({ error:undefined }))
      this.props.onSubmit({
        description:this.state.description,
        amount: parseFloat(this.state.amount,10),
        createdAt: this.state.createdAt.valueOf(),
        note:this.state.note
      })
    }
  }
  render() {
    return (
      <div>
        <h1>Add Expense</h1>
        <form onSubmit={this.onSubmit}>
          <input
            type="text" autoFocus placeholder="Description"
            value={this.state.description}
            onChange={this.onDescriptionChange}
          />
          <input
            type="text" placeholder="Amount"
            value={this.state.amount}
            onChange={this.onAmountChange}/>
          <SingleDatePicker
            date={this.state.createdAt}
            onDateChange={this.onDateChange}
            focused={this.state.calenderFocused}
            onFocusChange={this.onFocusChange}
            numberOfMonths={1}
            isOutsideRange={_=>false}
          />
          <textarea
            placeholder="Add a note for your expense (option)"
            value={this.state.note}
            onChange={this.onNoteChange}
          ></textarea>
          <button>Add Expense</button>
        </form>
        {this.state.error && <p>{this.state.error}</p>}
      </div>
    )
  }
}

// const ExpenseForm = ({dispatch}) => (
//   <div>
//     <form
//       onSubmit={e=>{
//         e.preventDefault()
//         dispatch(addExpense({
//           description:e.target.elements[0].value,
//           amount:e.target.elements[1].value
//           }))
//         console.log("created:",e.target.elements[0].value)
//         e.target.elements[0].value = ""
//         e.target.elements[1].value = ""
//         }
//       }
//     >
//       <label>Des:
//         <input type="text" placeholder="Description" autoFocus name="Des"/>
//         <input type="text" placeholder="Amount" autoFocus name="Amo"/>
//       </label>
//         <input type="submit" value="submit"/>
//     </form>
//   </div>
// )

// export default connect()(ExpenseForm)