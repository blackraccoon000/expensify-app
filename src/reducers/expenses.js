const expensesReducerDefaultState = []

const expensesReducer = (state = expensesReducerDefaultState,action) => {
  const { type, expense, id:actionId, updates } = action
  switch (type) {
    case "ADD_EXPENSE":
      return [
        ...state,
        expense
      ]
    case "REMOVE_EXPENSE":
      return state.filter(({id}) => id !== actionId)
    case "EDIT_EXPENSE":
      return state.map((expense) => {
        if(expense.id === actionId) {
          return {
            ...expense,
            ...updates
          }
        } else {
          return expense
        }
      })
    default:
      return state
  }
}

export default expensesReducer