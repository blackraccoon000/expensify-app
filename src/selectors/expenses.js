import moment from "moment"

// --> Ex: getVisibleExpenses(expenses,filters)
const getVisibleExpenses = (expenses, { text, sortBy, startDate, endDate }) => {
  const filterExpenses = expenses.filter(expense => {
    // const startDateMatch = typeof startDate !== "number" || expense.createdAt >= startDate
    // const endDateMatch = typeof endDate !== "number" || expense.createdAt <= endDate
    const createdAtMoment = moment(expense.createdAt)
    const startDateMatch = startDate ? startDate.isSameOrBefore(createdAtMoment , "day") : true
    const endDateMatch = endDate ? endDate.isSameOrAfter(createdAtMoment , "day") : true
    const textMatch = expense.description.toLowerCase().includes(text.toLowerCase())
    return startDateMatch && endDateMatch && textMatch
  })
  const sortExpenses = filterExpenses.sort((a,b) => {
    if(sortBy === "date") {
      return a.createdAt < b.createdAt ? 1 : -1
    } else if (sortBy === "amount") {
      return a.amount < b.amount ? 1 : -1
    }
  })
  return sortExpenses
}

export default getVisibleExpenses