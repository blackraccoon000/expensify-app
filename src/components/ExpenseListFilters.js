import React from "react"
import { connect } from "react-redux"
import { DateRangePicker } from "react-dates"
import { setTextFilter,sortByAmount,sortByDate, setStartDate,setEndDate } from "../actions/filters"

export class ExpenseListFilters extends React.Component {
  constructor(props){
    super(props)
    this.state = {
      calendarFocused:null
    }
  }
  onDatesChange = ({startDate, endDate}) => {
    this.props.startDate(startDate)
    this.props.endDate(endDate)
  }
  onFocusChange = ( calendarFocused ) => {
    this.setState({ calendarFocused })
  }
  onSetTextFilter = (e) => {
    this.props.setTextFilter(e.target.value)
  }
  onSortChange = (e) => {
    if(e.target.value==="date") {
      this.props.sortByDate()
    } else if (e.target.value==="amount") {
      this.props.sortByAmount()
    }
  }
  render() {
    const {filters} = this.props
    return (
      <div>
      <input type="text" value={filters.text}
        onChange = { this.onSetTextFilter }
      />
      <select
        value={filters.sortBy}
        onChange={this.onSortChange}>
        <option value="date">Date</option>
        <option value="amount">Amount</option>
      </select>
      <DateRangePicker
        startDate={filters.startDate}
        endDate={filters.endDate}
        onDatesChange={this.onDatesChange}
        focusedInput={this.state.calendarFocused}
        onFocusChange={this.onFocusChange}
        numberOfMonths={1} // 1月ごと表示
        isOutsideRange={()=>false} // 範囲外表示
        showClearDates={true} //　☓ボタン
      />
    </div>
    )
  }
}

const mapStateToProps = (state) => ({
  filters: state.filters
})

const mapDispatchToProps = (dispatch) => ({
  startDate: startDate => dispatch(setStartDate(startDate)),
  endDate: endDate => dispatch(setEndDate(endDate)),
  setTextFilter: text => dispatch(setTextFilter(text)),
  sortByDate: () => dispatch(sortByDate()),
  sortByAmount: () => dispatch(sortByAmount())
})

export default connect(mapStateToProps,mapDispatchToProps)(ExpenseListFilters)