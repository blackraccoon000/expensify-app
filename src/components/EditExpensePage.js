import React from "react"
import { connect } from "react-redux"
import { editExpense,startRemoveExpense } from "../actions/expenses"
import ExpenseForm from "./ExpenseForm"

export class EditExpensePage extends React.Component {
  constructor(props) {
    super(props)
  }
  onRemove= _ =>{
    this.props.startRemoveExpense({id:this.props.expense.id})
    this.props.history.push("/")
  }
  onSubmit= updates =>{
    this.props.editExpense(this.props.expense.id,updates)
    this.props.history.push("/")
  }
  render() {
    return (
      <div>
        <ExpenseForm
          expense={this.props.expense}
          onSubmit={this.onSubmit}
        />
        <button
          onClick={this.onRemove}
        >remove</button>
      </div>
    )
  }
}

const mapStateToProps = (state,props) => ({
  expense: state.expenses.find(expense => expense.id === props.match.params.id)
})

const mapDispatchToProps = (dispatch,props) => ({
  editExpense: (id,updates) => dispatch(editExpense(id,updates)),
  startRemoveExpense: id => dispatch(startRemoveExpense(id))
})

export default connect(mapStateToProps,mapDispatchToProps)(EditExpensePage)