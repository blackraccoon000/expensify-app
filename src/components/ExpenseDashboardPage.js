import React from "react"
import ExpenseList from "./ExpenseList"
import ExpenseListFilters from "./ExpenseListFilters"

const ExpenseDashboardPage = () => {
  return (
  <div>
    This is from my dashboard component
    <ExpenseListFilters />
    <ExpenseList />
  </div>
)}

export default ExpenseDashboardPage