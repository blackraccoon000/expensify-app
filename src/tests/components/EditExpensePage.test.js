import React from "react"
import { shallow } from "enzyme"
import { EditExpensePage } from "../../components/EditExpensePage"
import expenses from "../fixtures/expenses"

describe("EditExpensePageのテストを実施する",()=>{
  let editExpense, removeExpense, history, wrapper

  beforeEach(()=>{
    editExpense = jest.fn()
    removeExpense = jest.fn()
    history = { push: jest.fn() }
    wrapper = shallow(<EditExpensePage
      editExpense = {editExpense}
      removeExpense = {removeExpense}
      expense = { expenses[0] }
      history = {history}
    />)
  })
  it("should render EditExpensePage",()=>{
    expect(wrapper).toMatchSnapshot()
  })

  it("should handle editExpense", () => {
    const updates = {
      note:"New Note value",
      description: "New Description value"
    }
    wrapper.find("ExpenseForm").prop("onSubmit")(updates)
    expect(history.push).toHaveBeenLastCalledWith("/")
    expect(editExpense).toHaveBeenLastCalledWith(expenses[0].id,updates)
  })

  it("should handle removeExpense", () => {
    // wrapper.find("button").prop("onClick")()
    wrapper.find("button").simulate("click")
    expect(history.push).toHaveBeenLastCalledWith("/")
    expect(removeExpense).toHaveBeenLastCalledWith({id:expenses[0].id})
  })
})