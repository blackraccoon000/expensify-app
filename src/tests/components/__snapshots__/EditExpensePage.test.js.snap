// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EditExpensePageのテストを実施する should render EditExpensePage 1`] = `
<div>
  <ExpenseForm
    expense={
      Object {
        "amount": 195,
        "createdAt": 0,
        "description": "Gum",
        "id": "1",
        "note": "",
      }
    }
    onSubmit={[Function]}
  />
  <button
    onClick={[Function]}
  >
    remove
  </button>
</div>
`;
